README.rst
========
Overview
========

.. start-badges

.. list-table::
    :stub-columns: 1

    * - tests
      - | |github-actions| |requires|
        | |coveralls| |codecov|
    * - package
      - | |version| |wheel| |supported-versions| |supported-implementations|
        | |commits-since|

.. |github-actions| image:: https://github.com/ionelmc/tox-wheel/actions/workflows/github-actions.yml/badge.svg
    :alt: GitHub Actions Build Status
    :target: https://github.com/ionelmc/tox-wheel/actions

.. |requires| image:: https://requires.io/github/ionelmc/tox-wheel/requirements.svg?branch=master
    :alt: Requirements Status
    :target: https://requires.io/github/ionelmc/tox-wheel/requirements/?branch=master

.. |coveralls| image:: https://coveralls.io/repos/ionelmc/tox-wheel/badge.svg?branch=master&service=github
    :alt: Coverage Status
    :target: https://coveralls.io/r/ionelmc/tox-wheel

.. |codecov| image:: https://codecov.io/gh/ionelmc/tox-wheel/branch/master/graphs/badge.svg?branch=master
    :alt: Coverage Status
    :target: https://codecov.io/github/ionelmc/tox-wheel

.. |version| image:: https://img.shields.io/pypi/v/tox-wheel.svg
    :alt: PyPI Package latest release
    :target: https://pypi.org/project/tox-wheel

.. |wheel| image:: https://img.shields.io/pypi/wheel/tox-wheel.svg
    :alt: PyPI Wheel
    :target: https://pypi.org/project/tox-wheel

.. |supported-versions| image:: https://img.shields.io/pypi/pyversions/tox-wheel.svg
    :alt: Supported versions
    :target: https://pypi.org/project/tox-wheel

.. |supported-implementations| image:: https://img.shields.io/pypi/implementation/tox-wheel.svg
    :alt: Supported implementations
    :target: https://pypi.org/project/tox-wheel

.. |commits-since| image:: https://img.shields.io/github/commits-since/ionelmc/tox-wheel/v1.0.0.svg
    :alt: Commits since latest release
    :target: https://github.com/ionelmc/tox-wheel/compare/v1.0.0...master



.. end-badges

A `tox <http://tox.readthedocs.org>`_ plugin that builds and installs wheels instead of sdist.

* Free software: BSD 2-Clause License

Installation
============

::

    pip install tox-wheel

You can also install the in-development version with::

    pip install https://github.com/ionelmc/tox-wheel/archive/master.zip


Documentation
=============

Enabling
--------

To enable either use ``tox --wheel`` or change your ``tox.ini`` if you always want the plugin to be enabled:

.. code-block:: ini

    [testenv]
    wheel = true

You can also use factors in ``tox.ini``:

.. code-block:: ini

    [tox]
    envlist = {py27,py35,py36,py37,py38,pypy,pypy3}-{cover,nocov}

    [testenv]
    wheel =
        cover: false
        nocov: true


Build configuration
-------------------

This plugin will build wheels for all the active environments. Note that building will be done in a batch before any testing starts
(in order to support ``tox --parallel`` mode).

If you can produce universal wheels you might want to configure the build env so that the wheel is only built once for all the envs:

.. code-block:: ini

    [testenv]
    wheel_build_env = build

    [testenv:build]

Note that you can also use ``wheel_build_env`` for situation where you have many environments for the same interpreter:

.. code-block:: ini

    [testenv:py38]
    ; regular testing

    [testenv:py38-extras]
    ; tests with optional dependencies
    wheel_build_env = py38

    [testenv:docs]
    ; docs building
    wheel_build_env = py38

The plugin cleans the build dir by default, in case you want to speed things further (at the risk of build caching problems)
you could use ``tox --wheel-dirty``.

You can also place this configuration in ``tox.ini`` but there will be a unpleasant surprise factor if you
ever hit the aforementioned build problems:

.. code-block:: ini

    [testenv]
    wheel_dirty = true

PEP517 support
--------------

If you have a custom ``[build-system] build-backend`` in your ``pyproject.toml`` you need to enable the PEP517 builder by
having this in your ``tox.ini``:

.. code-block:: ini

    [testenv]
    wheel_pep517 = true

Enabling this will delegate building to ``pip wheel --use-pep517``.

Development
===========

To run the all tests run::

    tox


CHANGELOG.rst

Changelog
=========

1.0.0 (2022-10-01)
------------------

* Added option to build wheels (and sdists) in an isolated environment using `build <https://github.com/pypa/build>`_.
  Contributed by Ben Rowland in `#17 <https://github.com/ionelmc/tox-wheel/pull/17>`_.

0.7.0 (2021-12-29)
------------------

* Fixed build dir not being cleaned up in PEP 517 mode.
  Contributed by Michael Rans in `#16 <https://github.com/ionelmc/tox-wheel/pull/16>`_.
* Switched CI from Travis/AppVeyor to GitHub Actions.

0.6.0 (2020-11-06)
------------------

* Fixed unnecessary build directory cleanup and removed dead code.
  Contributed by Thomas Grainger in `#9 <https://github.com/ionelmc/tox-wheel/pull/9>`_.
* The ``isolated_build`` tox option is now an alias for ``wheel_pep517``.
  Contributed by Thomas Grainger in `#6 <https://github.com/ionelmc/tox-wheel/pull/6>`_.
* Added more configuration examples.

0.5.0 (2020-08-06)
------------------

* Added support for PEP 517/518.
  Contributed by Antonio Botelho in `#5 <https://github.com/ionelmc/tox-wheel/pull/5>`_.

0.4.2 (2019-05-15)
------------------

* Improved logging a bit so messages are less confusing (don't emit ``wheel-make path/to/setup.py``).
* Moved dist/build cleanup right before ``bdist_wheel``.

0.4.1 (2019-05-15)
------------------

* Improved error handling when no dists are built.

0.4.0 (2019-05-05)
------------------

* Fixed compatibility with tox and changed requirement for minimum tox version to 3.9.0.

0.3.0 (2019-01-26)
------------------

* Added support for ``tox --parallel`` mode.
* Added ``wheel_build_env`` config option.
* Renamed ``wheel_clean_build`` config option to ``wheel_dirty``.
* Added ``--wheel-dirty`` CLI argument.

0.2.1 (2019-01-12)
------------------

* Added ``wheel`` to dependencies.

0.2.0 (2019-01-12)
------------------

* Remove ``--wheel-clean-build`` CLI option. Build directory cleaning is now on by default.
  Correct behavior should be the default.
* Added support for ``[testenv] wheel`` (default: ``false``) and ``[testenv] wheel_clean_build`` (default: ``true``)
  in ``tox.ini``.

0.1.0 (2019-01-09)
------------------

* First release on PyPI.


tests/test_tox_wheel.py
import sys

import pytest

import tox_wheel.plugin

try:
    from unittest.mock import MagicMock
    from unittest.mock import patch
except ImportError:
    from mock import MagicMock
    from mock import patch

pytest_plugins = 'pytester',


@pytest.fixture
def testdir_legacy(testdir):
    testdir.tmpdir.join('tox.ini').write("""
[tox]
envlist = py-{a,b}
""")
    testdir.tmpdir.join('setup.py').write("""
from setuptools import setup

setup(name='foobar')
""")
    testdir.tmpdir.join('build').ensure(dir=1)
    return testdir


@pytest.fixture
def testdir_pep517(testdir):
    testdir.tmpdir.join('tox.ini').write("""
[tox]
envlist = py-{a,b}

[testenv]
wheel = true
wheel_pep517 = true
""")
    testdir.tmpdir.join('setup.py').write("""
from setuptools import setup

setup(name='foobar')
""")
    testdir.tmpdir.join('pyproject.toml').write("""
[build-system]
requires = [
    "setuptools >= 35.0.2"
]
build-backend = "setuptools.build_meta"
""")
    testdir.tmpdir.join('build').ensure(dir=1)
    return testdir


@pytest.fixture
def testdir_pep517_build(testdir):
    testdir.tmpdir.join('tox.ini').write("""
[tox]
envlist = py-{a,b}

[testenv]
wheel = true
wheel_pep517 = build
""")
    testdir.tmpdir.join('setup.py').write("""
from setuptools import setup

setup(name='foobar', packages=[])
""")
    testdir.tmpdir.join('pyproject.toml').write("""
[build-system]
requires = [
    "setuptools >= 35.0.2"
]
build-backend = "setuptools.build_meta"
""")
    testdir.tmpdir.join('build').ensure(dir=1)
    return testdir


@pytest.fixture(params=['', '--parallel 1 --parallel-live'], ids=['sequential', 'parallel'])
def options(request):
    return ['-e', 'py-a,py-b'] + request.param.split()


def test_patch():
    class A(object):
        def __init__(self, a):
            self.a = a

    obj = A(10)
    with tox_wheel.plugin.patch(obj, 'a', 5):
        assert obj.a == 5
    assert obj.a == 10


def test_disabled(testdir_legacy, options):
    result = testdir_legacy.run('tox', *options)
    result.stdout.fnmatch_lines([
        'GLOB sdist-make: *',
    ])


def test_enabled_legacy(testdir_legacy, options):
    result = testdir_legacy.run('tox', '--wheel', *options)
    result.stdout.fnmatch_lines([
        'py* wheel-make: *',
    ])
    assert result.stdout.str().count('running bdist_wheel') == 2
    assert result.ret == 0


def test_enabled_pep517(testdir_pep517, options):
    result = testdir_pep517.run('tox', *options)
    result.stdout.fnmatch_lines([
        'py* wheel-make: *',
    ])
    if sys.version_info >= (3, 6):
        build_string = 'Building wheel for foobar (pyproject.toml)'
    else:
        build_string = 'Building wheel for foobar (PEP 517)'
    assert result.stdout.str().count(build_string) == 4
    assert result.ret == 0


def test_enabled_pep517_build(testdir_pep517_build, options):
    result = testdir_pep517_build.run('tox', *options)
    result.stdout.fnmatch_lines([
        'py* wheel-make: *',
    ])
    build_string = 'Successfully built foobar-0.0.0.tar.gz and foobar-0.0.0-py3-none-any.whl'
    assert result.stdout.str().count(build_string) == 2
    assert result.ret == 0


def test_build_env_legacy(testdir_legacy, options):
    testdir_legacy.tmpdir.join('setup.cfg').write("""
[bdist_wheel]
universal = 1
""")
    testdir_legacy.tmpdir.join('tox.ini').write("""
[testenv]
wheel = true
wheel_build_env = build

[testenv:build]
""", mode='a')
    result = testdir_legacy.run('tox', *options)
    result.stdout.fnmatch_lines([
        'build wheel-make: *',
    ])
    assert result.stdout.str().count('running bdist_wheel') == 1
    assert result.ret == 0


def test_build_env_pep517(testdir_pep517, options):
    testdir_pep517.tmpdir.join('setup.cfg').write("""
[bdist_wheel]
universal = 1
""")
    testdir_pep517.tmpdir.join('tox.ini').write("""
wheel_build_env = build

[testenv:build]
""", mode='a')
    result = testdir_pep517.run('tox', *options)
    result.stdout.fnmatch_lines([
        'build wheel-make: *',
    ])
    if sys.version_info >= (3, 6):
        build_string = 'Building wheel for foobar (pyproject.toml)'
    else:
        build_string = 'Building wheel for foobar (PEP 517)'
    assert result.stdout.str().count(build_string) == 2
    assert result.ret == 0


def test_build_env_pep517_build(testdir_pep517_build, options):
    testdir_pep517_build.tmpdir.join('setup.cfg').write("""
[bdist_wheel]
universal = 1
""")
    testdir_pep517_build.tmpdir.join('tox.ini').write("""
wheel_build_env = build

[testenv:build]
""", mode='a')
    result = testdir_pep517_build.run('tox', *options)
    result.stdout.fnmatch_lines([
        'build wheel-make: *',
    ])
    build_string = 'Successfully built foobar-0.0.0.tar.gz and foobar-0.0.0-py2.py3-none-any.whl'
    assert result.stdout.str().count(build_string) == 1
    assert result.ret == 0


@pytest.mark.parametrize('wheel_build_env', ['', 'wheel_build_env'])
def test_skip_usedevelop(testdir_legacy, options, wheel_build_env):
    testdir_legacy.tmpdir.join('tox.ini').write("""
[testenv]
usedevelop = true
""" + ("""
wheel_build_env = build

[testenv:build]
""" if wheel_build_env else ""), mode='a')
    result = testdir_legacy.run('tox', '-v', '--wheel', *options)
    stdout = result.stdout.str()
    assert stdout.count('wheel-make') == 0
    assert stdout.count('bdist_wheel') == 0
    assert result.ret == 0


def test_enabled_toxini_noclean_legacy(testdir_legacy, options):
    testdir_legacy.tmpdir.join('tox.ini').write("""
[testenv]
wheel = true
wheel_dirty = true
""", mode='a')
    result = testdir_legacy.run('tox', *options)
    result.stdout.fnmatch_lines([
        'py* wheel-make: *',
    ])
    assert 'cleaning up build directory ...' not in result.stdout.str()
    assert 'cleaning up build directory ...' not in result.stderr.str()
    assert result.ret == 0


def test_enabled_toxini_noclean_pep517(testdir_pep517, options):
    testdir_pep517.tmpdir.join('tox.ini').write("""
wheel_dirty = true
""", mode='a')
    result = testdir_pep517.run('tox', *options)
    result.stdout.fnmatch_lines([
        'py* wheel-make: *',
    ])
    assert 'cleaning up build directory ...' not in result.stdout.str()
    assert 'cleaning up build directory ...' not in result.stderr.str()
    assert result.ret == 0


def test_enabled_cli_noclean(testdir_legacy, options):
    testdir_legacy.tmpdir.join('tox.ini').write("""
[testenv]
wheel = true
""", mode='a')
    result = testdir_legacy.run('tox', '--wheel-dirty', *options)
    result.stdout.fnmatch_lines([
        'py* wheel-make: *',
    ])
    assert 'cleaning up build directory ...' not in result.stdout.str()
    assert 'cleaning up build directory ...' not in result.stderr.str()
    assert result.ret == 0


def test_enabled_toxini(testdir_legacy, options):
    testdir_legacy.tmpdir.join('tox.ini').write("""
[testenv]
wheel = true
""", mode='a')
    result = testdir_legacy.run('tox', '-vv', *options)
    result.stdout.fnmatch_lines([
        'py* wheel-make: *',
        'py* wheel-make: cleaning up build directory ...',
        '  removing *[\\/]build',
        'py* finish: packaging *',
        'copying new sdistfile to *.whl*',
    ])
    assert 'is not a supported wheel on this platform.' not in result.stdout.str()
    assert 'is not a supported wheel on this platform.' not in result.stderr.str()


def test_skip_missing_interpreters():
    with patch.object(tox_wheel.plugin, 'get_package') as mock_build:
        venv = MagicMock()
        venv.envconfig.wheel = True
        session = MagicMock()
        session.config.option.wheel = True
        session.getvenv.return_value = object()
        mock_build.side_effect = tox_wheel.plugin.InterpreterNotFound("No interpreter")

        with pytest.raises(tox_wheel.plugin.InterpreterNotFound):
            tox_wheel.plugin.get_package(session)

        session.config.option.skip_missing_interpreters = True
        assert tox_wheel.plugin.tox_package(session, venv) is None

        session.config.option.skip_missing_interpreters = False
        with pytest.raises(tox_wheel.plugin.InterpreterNotFound):
            tox_wheel.plugin.tox_package(session, venv)


def test_multiplex_sdist_and_wheel(testdir_legacy, options):
    testdir_legacy.tmpdir.join('tox.ini').write("""
[tox]
envlist =
    py-{a,b}
    missing_interpreter

skip_missing_interpreters = True

[testenv:a]
wheel = false

[testenv:b]
wheel = true

[testenv:missing_interpreter]
basepython = python3.nothing
""")
    assert 'py-a,py-b' in options
    options[options.index('py-a,py-b')] = 'py-a,py-b,missing_interpreter'

    result = testdir_legacy.run('tox', '-vv', *options)
    result.stdout.fnmatch_lines([
        'GLOB sdist-make: *',
        '*Building wheels*',
    ])
    assert result.ret == 0, result.stdout


ci/appveyor-download.py
#!/usr/bin/env python
"""
Use the AppVeyor API to download Windows artifacts.

Taken from: https://bitbucket.org/ned/coveragepy/src/tip/ci/download_appveyor.py
# Licensed under the Apache License: http://www.apache.org/licenses/LICENSE-2.0
# For details: https://bitbucket.org/ned/coveragepy/src/default/NOTICE.txt
"""
from __future__ import unicode_literals

import argparse
import os
import zipfile

import requests


def make_auth_headers():
    """Make the authentication headers needed to use the Appveyor API."""
    path = os.path.expanduser("~/.appveyor.token")
    if not os.path.exists(path):
        raise RuntimeError(
            "Please create a file named `.appveyor.token` in your home directory. "
            "You can get the token from https://ci.appveyor.com/api-token"
        )
    with open(path) as f:
        token = f.read().strip()

    headers = {
        'Authorization': 'Bearer {}'.format(token),
    }
    return headers


def download_latest_artifacts(account_project, build_id):
    """Download all the artifacts from the latest build."""
    if build_id is None:
        url = "https://ci.appveyor.com/api/projects/{}".format(account_project)
    else:
        url = "https://ci.appveyor.com/api/projects/{}/build/{}".format(account_project, build_id)
    build = requests.get(url, headers=make_auth_headers()).json()
    jobs = build['build']['jobs']
    print(u"Build {0[build][version]}, {1} jobs: {0[build][message]}".format(build, len(jobs)))

    for job in jobs:
        name = job['name']
        print(u"  {0}: {1[status]}, {1[artifactsCount]} artifacts".format(name, job))

        url = "https://ci.appveyor.com/api/buildjobs/{}/artifacts".format(job['jobId'])
        response = requests.get(url, headers=make_auth_headers())
        artifacts = response.json()

        for artifact in artifacts:
            is_zip = artifact['type'] == "Zip"
            filename = artifact['fileName']
            print(u"    {0}, {1} bytes".format(filename, artifact['size']))

            url = "https://ci.appveyor.com/api/buildjobs/{}/artifacts/{}".format(job['jobId'], filename)
            download_url(url, filename, make_auth_headers())

            if is_zip:
                unpack_zipfile(filename)
                os.remove(filename)


def ensure_dirs(filename):
    """Make sure the directories exist for `filename`."""
    dirname = os.path.dirname(filename)
    if dirname and not os.path.exists(dirname):
        os.makedirs(dirname)


def download_url(url, filename, headers):
    """Download a file from `url` to `filename`."""
    ensure_dirs(filename)
    response = requests.get(url, headers=headers, stream=True)
    if response.status_code == 200:
        with open(filename, 'wb') as f:
            for chunk in response.iter_content(16 * 1024):
                f.write(chunk)
    else:
        print(u"    Error downloading {}: {}".format(url, response))


def unpack_zipfile(filename):
    """Unpack a zipfile, using the names in the zip."""
    with open(filename, 'rb') as fzip:
        z = zipfile.ZipFile(fzip)
        for name in z.namelist():
            print(u"      extracting {}".format(name))
            ensure_dirs(name)
            z.extract(name)


parser = argparse.ArgumentParser(description='Download artifacts from AppVeyor.')
parser.add_argument('--id',
                    metavar='PROJECT_ID',
                    default='ionelmc/tox-wheel',
                    help='Project ID in AppVeyor.')
parser.add_argument('build',
                    nargs='?',
                    metavar='BUILD_ID',
                    help='Build ID in AppVeyor. Eg: master-123')

if __name__ == "__main__":
    # import logging
    # logging.basicConfig(level="DEBUG")
    args = parser.parse_args()
    download_latest_artifacts(args.id, args.build)


src/tox_wheel/plugin.py
import os.path
from contextlib import contextmanager
from functools import partial

import pluggy
import py
from tox import package
from tox import reporter
from tox.exception import InterpreterNotFound
from tox.package import get_package
from tox.util.path import ensure_empty_dir

hookimpl = pluggy.HookimplMarker("tox")


@hookimpl
def tox_addoption(parser):
    parser.add_argument(
        "--wheel",
        action="store_true",
        help="Build wheel instead of sdist",
    )
    parser.add_argument(
        "--wheel-pep517",
        action="store_true",
        help="Build wheel using PEP 517/518"
    )
    parser.add_argument(
        "--wheel-dirty",
        action="store_true",
        help="Do not remove build directory (fast but dirty builds)",
    )
    parser.add_testenv_attribute(
        name="wheel",
        type="bool",
        default=False,
        help="Build wheel instead of sdist",
    )
    parser.add_testenv_attribute(
        name="wheel_pep517",
        type="string",
        default="",
        help=(
            "Build wheel using PEP 517/518 (pass true to build with pip or "
            "build to build with build)"
        ),
    )
    parser.add_testenv_attribute(
        name="wheel_dirty",
        type="bool",
        default=False,
        help="Do not remove build directory (fast but dirty builds)"
    )
    parser.add_testenv_attribute(
        name="wheel_build_env",
        type="string",
        default="{envname}",
        help="Environment to use for building the wheel. Default: %(default)r"
    )


@contextmanager
def patch(obj, attr, value):
    original = getattr(obj, attr)
    setattr(obj, attr, value)
    try:
        yield
    finally:
        setattr(obj, attr, original)


@hookimpl
def tox_testenv_install_deps(venv, action):
    if venv.envconfig.wheel_pep517 == "build":
        venv.run_install_command(["build[virtualenv]>=0.7.0"], action)
    return None


@hookimpl
def tox_package(session, venv):
    if session.config.option.wheel or venv.envconfig.wheel:
        build_venv = session.getvenv(venv.envconfig.wheel_build_env)
        if not hasattr(build_venv, "wheel_package"):
            try:
                with patch(package, "build_package", partial(wheel_build_package, venv=build_venv)):
                    build_venv.wheel_package, build_venv.wheel_dist = get_package(session)
            except InterpreterNotFound:
                if session.config.option.skip_missing_interpreters:
                    return None
                raise
        return build_venv.wheel_package

    elif hasattr(session, "package"):
        return session.package


def wheel_build_package(config, session, venv):
    if config.isolated_build or session.config.option.wheel_pep517 or venv.envconfig.wheel_pep517:
        wheel_package = wheel_build_pep517(config, session, venv)
    else:
        wheel_package = wheel_build_legacy(config, session, venv)
    return wheel_package


def wheel_is_allowed_external(path, venv, is_allowed_external=None):
    if is_allowed_external is None:
        is_allowed_external = venv.is_allowed_external
    if not is_allowed_external(path):
        raise RuntimeError("Couldn't find interpreter inside {} for building".format(venv))
    return True


def wheel_build_legacy(config, session, venv):
    setup = config.setupdir.join("setup.py")
    if not setup.check():
        reporter.error("No setup.py file found. The expected location is: {}".format(setup))
        raise SystemExit(1)
    with session.newaction(venv.name, "packaging") as action:
        with patch(venv, "is_allowed_external", partial(wheel_is_allowed_external, venv=venv)):
            venv.update(action=action)
            if not (session.config.option.wheel_dirty or venv.envconfig.wheel_dirty):
                action.setactivity("wheel-make", "cleaning up build directory ...")
                ensure_empty_dir(config.setupdir.join("build"))
            ensure_empty_dir(config.distdir)
            venv.test(
                name="wheel-make",
                commands=[["python", setup, "bdist_wheel", "--dist-dir", config.distdir]],
                redirect=False,
                ignore_outcome=False,
                ignore_errors=False,
                display_hash_seed=False,
            )
        try:
            dists = config.distdir.listdir()
        except py.error.ENOENT:
            # check if empty or comment only
            data = []
            with open(str(setup)) as fp:
                for line in fp:
                    if line and line[0] == "#":
                        continue
                    data.append(line)
            if not "".join(data).strip():
                reporter.error("setup.py is empty")
                raise SystemExit(1)
            reporter.error(
                "No dist directory found. Please check setup.py, e.g with:\n"
                "     python setup.py bdist_wheel"
            )
            raise SystemExit(1)
        else:
            if not dists:
                reporter.error(
                    "No distributions found in the dist directory found. Please check setup.py, e.g with:\n"
                    "     python setup.py bdist_wheel"
                )
                raise SystemExit(1)
            return dists[0]


def wheel_build_pep517(config, session, venv):
    pyproject = config.setupdir.join("pyproject.toml")
    if not pyproject.check():
        reporter.error("No pyproject.toml file found. The expected location is: {}".format(pyproject))
        raise SystemExit(1)
    with session.newaction(venv.name, "packaging") as action:
        venv.update(action=action)
        if not (session.config.option.wheel_dirty or venv.envconfig.wheel_dirty):
            action.setactivity("wheel-make", "cleaning up build directory ...")
            ensure_empty_dir(config.setupdir.join("build"))
        ensure_empty_dir(config.distdir)
        if venv.envconfig.wheel_pep517 == "build":
            commands = [
                "python",
                "-Im",
                "build",
                "--outdir",
                config.distdir,
                config.setupdir,
            ]
        else:
            commands = [
                "pip",
                "wheel",
                config.setupdir,
                "--no-deps",
                "--use-pep517",
                "--wheel-dir",
                config.distdir,
            ]
        venv.test(
            name="wheel-make",
            commands=[commands],
            redirect=False,
            ignore_outcome=False,
            ignore_errors=False,
            display_hash_seed=False,
        )
        try:
            # we need to filter our list of dists to include only wheels
            dists = [dist for dist in config.distdir.listdir() if os.path.splitext(dist)[1] == ".whl"]
        except py.error.ENOENT:
            reporter.error(
                "No dist directory found. Please check pyproject.toml, e.g with:\n"
                "     pip wheel . --use-pep517"
            )
            raise SystemExit(1)
        else:
            if not dists:
                reporter.error(
                    "No distributions found in the dist directory found. Please check pyproject.toml, e.g with:\n"
                    "     pip wheel . --use-pep517"
                )
                raise SystemExit(1)
            return dists[0]
